File: src/features/typing/data/go/largest_rectangle_in_histogram.go
package main

func largestRectangleArea(heights []int) int {
	stack := []int{}
	best := 0
	popArea := func(right int) {
		h := heights[stack[len(stack)-1]]
		stack = stack[:len(stack)-1]
		width := right
		if len(stack) > 0 {
			width = right - stack[len(stack)-1] - 1
		}
		if h*width > best {
			best = h * width
		}
	}
	for i := 0; i < len(heights); i++ {
		for len(stack) > 0 && heights[stack[len(stack)-1]] >= heights[i] {
			popArea(i)
		}
		stack = append(stack, i)
	}
	for len(stack) > 0 {
		popArea(len(heights))
	}
	return best
}

File: src/features/typing/data/ts/largest_rectangle_in_histogram.ts
function largestRectangleArea(heights: number[]): number {
  const stack: number[] = [];
  let best = 0;
  const popArea = (right: number): void => {
    const h = heights[stack[stack.length - 1]];
    stack.pop();
    const width = stack.length === 0 ? right : right - stack[stack.length - 1] - 1;
    best = Math.max(best, h * width);
  };
  for (let i = 0; i < heights.length; i++) {
    while (stack.length > 0 && heights[stack[stack.length - 1]] >= heights[i]) {
      popArea(i);
    }
    stack.push(i);
  }
  while (stack.length > 0) {
    popArea(heights.length);
  }
  return best;
}
